In [12]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
from sklearn.cluster import KMeans

In [13]:
df = pd.read_csv('data_preprocessed.csv')
df

,Block,Description,Location Description,Arrest,Domestic,Beat,District,Ward,Community Area,Year,...,Crime_OTHER NARCOTIC VIOLATION,Crime_OTHER OFFENSE,Crime_PROSTITUTION,Crime_PUBLIC PEACE VIOLATION,Crime_RITUALISM,Crime_ROBBERY,Crime_SEX OFFENSE,Crime_STALKING,Crime_THEFT,Crime_WEAPONS VIOLATION
0,059XX S INDIANA AV,POSS: CANNABIS MORE THAN 30GMS,STREET,1.0,0.0,233.0,2.0,42.0,25.0,2001.0,...,False,False,False,False,False,False,False,False,False,False
1,027XX N LAWNDALE AV,SIMPLE,RESIDENCE,0.0,0.0,2524.0,25.0,42.0,25.0,2001.0,...,False,False,False,False,False,False,False,False,False,False
2,048XX W FULLERTON AV,HARASSMENT BY TELEPHONE,DRUG STORE,0.0,0.0,2521.0,25.0,42.0,25.0,2001.0,...,False,True,False,False,False,False,False,False,False,False
3,052XX W DIVISION ST,HARASSMENT BY TELEPHONE,STREET,0.0,0.0,2532.0,25.0,42.0,25.0,2001.0,...,False,True,False,False,False,False,False,False,False,False
4,022XX S KEDVALE AV,HARASSMENT BY TELEPHONE,RESIDENCE,0.0,1.0,1013.0,10.0,42.0,25.0,2001.0,...,False,True,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
52020,003XX W 103RD ST,SIMPLE,DEPARTMENT STORE,0.0,0.0,511.0,5.0,9.0,49.0,2015.0,...,False,False,False,False,False,False,False,False,False,False
52021,010XX N LATROBE AVE,DOMESTIC BATTERY SIMPLE,SIDEWALK,0.0,1.0,1524.0,15.0,37.0,25.0,2015.0,...,False,False,False,False,False,False,False,False,False,False
52022,031XX S PULASKI RD,SIMPLE,PARKING LOT/GARAGE(NON.RESID.),1.0,0.0,1031.0,10.0,22.0,30.0,2015.0,...,False,False,False,False,False,False,False,False,False,False
52023,066XX N ASHLAND AVE,TO PROPERTY,RESIDENCE,1.0,0.0,2432.0,24.0,40.0,1.0,2015.0,...,False,False,False,False,False,False,False,False,False,False


In [14]:
bool_cols = df.select_dtypes(include=["bool"]).columns
for col in bool_cols:
    df[col] = df[col].apply(lambda x: 0.0 if x is False else 1.0)


numeric_cols = df.select_dtypes(include=["int64","float64"]).columns
X = df[numeric_cols]

print(list(numeric_cols))

['Arrest', 'Domestic', 'Beat', 'District', 'Ward', 'Community Area', 'Year', 'Latitude', 'Longitude', 'Month', 'DayOfWeek', 'Hour', 'Loc_Commercial', 'Loc_Institution', 'Loc_Other', 'Loc_Public', 'Loc_Residential', 'Loc_Transportation', 'Crime_ARSON', 'Crime_ASSAULT', 'Crime_BATTERY', 'Crime_BURGLARY', 'Crime_CONCEALED CARRY LICENSE VIOLATION', 'Crime_CRIM SEXUAL ASSAULT', 'Crime_CRIMINAL DAMAGE', 'Crime_CRIMINAL SEXUAL ASSAULT', 'Crime_CRIMINAL TRESPASS', 'Crime_DECEPTIVE PRACTICE', 'Crime_GAMBLING', 'Crime_HOMICIDE', 'Crime_HUMAN TRAFFICKING', 'Crime_INTERFERENCE WITH PUBLIC OFFICER', 'Crime_INTIMIDATION', 'Crime_KIDNAPPING', 'Crime_LIQUOR LAW VIOLATION', 'Crime_MOTOR VEHICLE THEFT', 'Crime_NARCOTICS', 'Crime_NON - CRIMINAL', 'Crime_NON-CRIMINAL', 'Crime_OBSCENITY', 'Crime_OFFENSE INVOLVING CHILDREN', 'Crime_OTHER NARCOTIC VIOLATION', 'Crime_OTHER OFFENSE', 'Crime_PROSTITUTION', 'Crime_PUBLIC PEACE VIOLATION', 'Crime_RITUALISM', 'Crime_ROBBERY', 'Crime_SEX OFFENSE', 'Crime_STALKING',

In [19]:
kmeans = KMeans(n_clusters=4, random_state=1)
df["cluster"] = kmeans.fit_predict(X)
df["cluster"]

0        0
1        3
2        3
3        3
4        2
        ..
52020    0
52021    1
52022    2
52023    3
52024    1
Name: cluster, Length: 52025, dtype: int32

In [18]:
cluster_counts = df["cluster"].value_counts().sort_index()
print("Cluster counts:\n", cluster_counts)


Cluster counts:
 cluster
0    14847
1    12590
2    16244
3     8344
Name: count, dtype: int64


In [28]:
df.groupby("cluster")[numeric_cols].mean()


,Arrest,Domestic,Beat,District,Ward,Community Area,Year,Latitude,Longitude,Month,...,Crime_OTHER NARCOTIC VIOLATION,Crime_OTHER OFFENSE,Crime_PROSTITUTION,Crime_PUBLIC PEACE VIOLATION,Crime_RITUALISM,Crime_ROBBERY,Crime_SEX OFFENSE,Crime_STALKING,Crime_THEFT,Crime_WEAPONS VIOLATION
cluster,,,,,,,,,,,,,,,,,,,,,
0,0.262410,0.163265,383.263083,3.617633,24.798343,37.981949,2009.164276,41.768830,-87.608728,5.378595,...,0.00000,0.070385,0.004378,0.005927,0.000000,0.040277,0.004647,0.000471,0.203610,0.011787
1,0.262351,0.111835,1652.500079,16.235187,38.314138,20.680778,2008.329786,41.923034,-87.705527,4.936934,...,0.00000,0.060604,0.014376,0.005957,0.000000,0.032566,0.006116,0.001033,0.264813,0.006116
2,0.309653,0.144607,957.460293,9.346590,28.326828,37.022778,2009.046356,41.826289,-87.691275,5.339141,...,0.00000,0.062300,0.012928,0.007880,0.000062,0.041554,0.005294,0.000677,0.166523,0.013236
3,0.294703,0.130034,2338.073706,20.996045,38.840125,28.686122,2007.190077,41.891083,-87.691194,4.350551,...,0.00012,0.083533,0.012464,0.006112,0.000000,0.035834,0.005633,0.000360,0.205896,0.009108


In [27]:
cluster_profiles = df.groupby("cluster")[numeric_cols].mean()
for cluster_id in cluster_profiles.index:
    print(f"\nTop features for Cluster {cluster_id}:")
    print(cluster_profiles.loc[cluster_id].sort_values(ascending=False))




Top features for Cluster 0:
Year                                       2009.164276
Beat                                        383.263083
Latitude                                     41.768830
Community Area                               37.981949
Ward                                         24.798343
Hour                                         13.230013
Month                                         5.378595
District                                      3.617633
DayOfWeek                                     3.043039
Loc_Residential                               0.408163
Loc_Public                                    0.351384
Arrest                                        0.262410
Crime_THEFT                                   0.203610
Crime_BATTERY                                 0.186570
Domestic                                      0.163265
Loc_Commercial                                0.119687
Crime_CRIMINAL DAMAGE                         0.105678
Crime_NARCOTICS                     

In [ ]:
with open("clusters.txt", "w") as f:
    for cluster_id in cluster_profiles.index:
        f.write(f"\nTop features for Cluster {cluster_id}:\n")
        sorted_features = cluster_profiles.loc[cluster_id].sort_values(ascending=False)
        for feature, value in sorted_features.items():
            f.write(f"{feature:40} {value:.6f}\n")



### Cluster 0
- **Timeframe**: Around 2009, crimes mostly in the early afternoon around (1 pm).
- **Location**: Primarily residential (41%) and public areas (35%), with some commercial presence.
- **Crimes**: Theft (20%), battery (19%), domestic incidents (16%), criminal damage (10%), narcotics (8%).
- **Arrest rate**: 26%.

#### **Conclusion**: Cluster 0 represents general urban crimes around 2009, dominated by theft and battery, often in residential/public spaces, with moderate arrests.



### Cluster 1
- **Timeframe**: Around 2008, slightly earlier than Cluster 0.
- **Location**: Public (38%) and residential (34%), with higher commercial presence (15%).
- **Crimes**: Theft (26%), battery (15%), narcotics (9%), criminal damage (10%).
- **Arrest rate**: 26%.

#### **Conclusion**: Cluster 1 reflects slightly older crimes (2008), with theft as the dominant offense, occurring in public/commercial areas, and a moderate arrest rate.



### Cluster 2
- **Timeframe**: Around 2009.
- **Location**: Public (43%) and residential (38%), with smaller commercial share (8%).
- **Crimes**: Battery (18%), theft (17%), narcotics (13%), domestic (14%), criminal damage (11%).
- **Arrest rate**: 31%

#### **Conclusion**: Cluster 2 represents 2009 crimes with a stronger mix of battery, narcotics, and domestic incidents, occurring mostly in public spaces, with a higher arrest rate.



### Cluster 3
- **Timeframe**: Around 2007.
- **Location**: Residential (40%) and public (38%), with commercial at 11%.
- **Crimes**: Theft (21%), battery (16%), narcotics (10%), criminal damage (11%), domestic (13%).
- **Arrest rate**: 29%

#### **Conclusion**: Cluster 3 represents older crimes (2007), dominated by theft and battery, with notable domestic and narcotics presence, and a relatively high arrest rate.

